python/cudf/cudf/core/dtypes.py

```diff
@@ -8,9 +8,344 @@
 from pandas.api.extensions import ExtensionDtype
 
 import cudf
+from cudf._lib.types import _Dtype
 
 
-class CategoricalDtype(ExtensionDtype):
+class Generic(ExtensionDtype, _Dtype):
+    pa_type = None
+
+    def __eq__(self, other):
+        if isinstance(other, self.__class__):
+            return True
+        if isinstance(other, Generic) and not isinstance(
+            other, self.__class__
+        ):
+            return False
+        if (
+            isinstance(other, self.to_pandas.__class__)
+            or other is self.to_pandas.__class__
+        ):
+            return True
+
+        if self.to_numpy == other:
+            return True
+        if isinstance(other, str) and str(self.to_numpy) == other:
+            return True
+        return False
+
+    def __str__(self):
+        return self.name
+
+    @property
+    def num(self):
+        return self.to_numpy.num
+
+    @property
+    def to_numpy(self):
```

Comment:
If we want these to be properties, then let's rename them to something like `numpy_dtype`. `to_numpy` sounds like a method rather than a property. Assuming Dtypes are immutable we could probably make them `cached_properties` too

Reply:
I like the idea of making them part of the definition of the individual dtype objects themselves. To mimic pandas here I'll also rename this to `numpy_dtype`. Pandas does use cached properties here.

```
>>> pd.Int64Dtype.numpy_dtype
<pandas._libs.properties.CachedProperty object at 0x7fdbd2e48190>
```

I like the idea of class attributes just because then we can use them to map between datatypes without having to instantiate the cudf dtype. So `cudf.Int64Dtype.numpy_dtype` would work if someone needed to use that. But to be consistent with them perhaps we should use cached properties for now.